### ISSUE TITLE
Extended bounds create misaligned empty buckets in date histogram aggregation.

### ISSUE THREAD

**Issue by miccon** [elasticsearch/elasticsearch#8209]
Using a timezone in combination with 'pre_zone_adjust_large_interval' set to true leads to the creation of buckets which are not correctly aligned. 

Example:
Timezone set to CET, interval set to month (march and april are off here):
````
"aggregations" : {
    "histo" : {
      "buckets" : [ {
        "key_as_string" : "2013-12-31T23:00:00.000Z",
        "key" : 1388530800000,
        "doc_count" : 1
      }, {
        "key_as_string" : "2014-01-31T23:00:00.000Z",
        "key" : 1391209200000,
        "doc_count" : 0
      }, {
        "key_as_string" : "2014-02-28T23:00:00.000Z",
        "key" : 1393628400000,
        "doc_count" : 0
      }, {
        "key_as_string" : "2014-03-28T23:00:00.000Z",
        "key" : 1396047600000,
        "doc_count" : 0
      }, {
        "key_as_string" : "2014-04-28T23:00:00.000Z",
        "key" : 1398726000000,
        "doc_count" : 0
      } ]
    }
  }
````

Reproduction:
https://gist.github.com/miccon/4eecfeafc3a66a9d8b24

**Comment by wojcikstefan** [comment]
+1

Although the title is misleading - the issue has nothing to do with extended bounds.

Running the commands below:

```
curl -sXDELETE 'localhost:9200/test'
curl -sXPOST 'localhost:9200/test/test/?pretty=true&refresh=true' -d '{"date": "2014-01-01T0:00:00Z"}'
curl -sXPOST 'localhost:9200/test/test/?pretty=true&refresh=true' -d '{"date": "2014-04-01T0:00:00Z"}'
curl -sXPOST 'localhost:9200/test/test/?pretty=true&refresh=true' -d '{"date": "2014-04-30T0:00:00Z"}'
curl -sXGET 'localhost:9200/_search?pretty=true' -d '                                                                                                                                                                                                                                                                                            
{
  "size": 0,
  "aggs": {
    "histo": {
      "date_histogram": {
        "field": "date",
        "interval": "month",
        "pre_zone": "+01:00",
        "pre_zone_adjust_large_interval": true,
        "min_doc_count": 0
      }
    }
  }
}'
```

Results in:

```
{
  "took" : 6,
  "timed_out" : false,
  "_shards" : {
    "total" : 127,
    "successful" : 127,
    "failed" : 0
  },
  "hits" : {
    "total" : 30155,
    "max_score" : 0.0,
    "hits" : [ ]
  },
  "aggregations" : {
    "histo" : {
      "buckets" : [ {
        "key_as_string" : "2013-12-31T23:00:00.000Z",
        "key" : 1388530800000,
        "doc_count" : 1
      }, {
        "key_as_string" : "2014-01-31T23:00:00.000Z",
        "key" : 1391209200000,
        "doc_count" : 0
      }, {
        "key_as_string" : "2014-02-28T23:00:00.000Z",
        "key" : 1393628400000,
        "doc_count" : 0
      }, {
        "key_as_string" : "2014-03-28T23:00:00.000Z",
        "key" : 1396047600000,
        "doc_count" : 0
      }, {
        "key_as_string" : "2014-03-31T23:00:00.000Z",
        "key" : 1396306800000,
        "doc_count" : 2
      } ]
    }
  }
}
```

The bucket keys I'd expect are: `2013-12-31T23:00:00.000Z`, `2014-01-31T23:00:00.000Z`, `2014-02-28T23:00:00.000Z`, `2014-03-31T23:00:00.000Z`. Not that this is only an issue with positive `pre_zone` values and `pre_zone_adjust_large_interval` equal to `true`.

**Comment by cbuescher** [comment]
This looks very similar to #9491 and #7673 to me. Probably also fixed the same way. Will see if this behaviour is already fixed with the latest clean up of the `date_histogram` on master.

**Comment by cbuescher** [comment]
I tried the script of @wojcikstefan on current master (7c20a8), works there. However, bug is reproducable on 1.x. Will dig into this further.

**Comment by cbuescher** [comment]
@jpountz Had a look at these cases and why they still don't work on 1.x an 1.4 even after fix from https://github.com/elasticsearch/elasticsearch/pull/9790. The reason is that using `pre_zone_adjust_large_interval = true` for month and bigger intervals forces the TimeZoneRounding implementation to be TimeTimeZoneRoundingFloor. When calculating next buckets keys when inserting empty buckets we should do the adding of the time duration in local time (where also the rounding takes place).
Adding the back and forth conversion for preTz in TimeTimeZoneRoundingFloor.nextRoundingValue solved the issue for me. Will issue a PR if you want to have a look.

**Comment by jpountz** [comment]
@cbuescher +1 on a PR, your description of the fix looks good to me. Thanks for taking care of this!

**Comment by cbuescher** [comment]
Fix on 1.4 branch: e869e90
Fix on 1.x branch: f829670
Only tests on master: 4ef430d
